src/GraphQL.php

```diff
@@ -185,15 +188,77 @@ public function executeQuery(
         $rootValue = null,
         ?string $operationName = null
     ): ExecutionResult {
-        // TODO make executeQuery require a DocumentNode and move this parsing out of here
         if (is_string($query)) {
-            try {
-                $query = Parser::parse($query);
-            } catch (SyntaxError $syntaxError) {
-                return new ExecutionResult(null, [$syntaxError]);
-            }
+            return $this->parseAndExecuteQuery($query, $context, $variables, $rootValue, $operationName);
+        }
+
+        return $this->executeParsedQuery($query, $context, $variables, $rootValue, $operationName);
+    }
+
+    /**
+     * Parses query and executes it.
+     *
+     * @param array<string, mixed>|null $variables
+     * @param mixed|null $rootValue
+     */
+    public function parseAndExecuteQuery(
+        string $query,
+        GraphQLContext $context,
+        ?array $variables = [],
+        $rootValue = null,
+        ?string $operationName = null
+    ): ExecutionResult {
+        try {
+            $parsedQuery = $this->parse($query);
+        } catch (SyntaxError $syntaxError) {
+            return new ExecutionResult(null, [$syntaxError]);
         }
 
+        return $this->executeParsedQuery($parsedQuery, $context, $variables, $rootValue, $operationName);
+    }
+
+    /**
+     * Parses string query to DocumentNode. Caches queries if required.
+     *
+     * @throws SyntaxError
+     */
+    public function parse(string $query): DocumentNode
+    {
+        $cacheConfig = $this->configRepository->get('lighthouse.query_cache');
+
+        if (! $cacheConfig['enable']) {
+            return Parser::parse($query);
+        }
+
+        /** @var \Illuminate\Contracts\Cache\Factory $cacheFactory */
+        $cacheFactory = app(CacheFactory::class);
+        $store = $cacheFactory->store($cacheConfig['store']);
+
+        return $store->remember(
+            'lighthouse:query:' . hash('sha256', $query),
```

Comment:
We don't use global helpers such as `hash()`, they are usually not compatible with Lumen.

Reply:
`hash()` is a global php function - https://www.php.net/manual/en/function.hash.php

Reply:
You are right, thanks!